File: 138.copy-list-with-random-pointer.go
/*
 * @lc app=leetcode id=138 lang=golang
 *
 * [138] Copy List with Random Pointer
 */

package main

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func _copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	// build interweaved list
	current := head
	for current != nil {
		node := Node{Val: current.Val, Next: current.Next}
		current.Next = &node
		current = current.Next.Next
	}

	// set random nodes
	current = head
	for current != nil {
		if current.Random != nil {
			current.Next.Random = current.Random.Next
		}
		current = current.Next.Next
	}

	// fix original array
	current = head
	newHead := head.Next
	for current != nil {
		newNode := current.Next
		current.Next = newNode.Next
		if newNode.Next == nil {
			newNode.Next = nil
		} else {
			newNode.Next = newNode.Next.Next
		}
		current = current.Next
	}

	return newHead
}

// @lc code=start
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

// this is slower
func copyRandomList(head *Node) *Node {
	if head == nil {
		return head
	}

	nodeMap := make(map[*Node]*Node)

	current := head
	for current != nil {
		nodeMap[current] = &Node{Val: current.Val}
		current = current.Next
	}

	current = head
	for current != nil {
		nodeMap[current].Random = nodeMap[current.Random]
		nodeMap[current].Next = nodeMap[current.Next]
		current = current.Next
	}

	return nodeMap[head]
}

// @lc code=end

File: 138.copy-list-with-random-pointer.ts
/*
 * @lc app=leetcode id=138 lang=typescript
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/**
 * Definition for Node.
 * class Node {
 *     val: number
 *     next: Node | null
 *     random: Node | null
 *     constructor(val?: number, next?: Node, random?: Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */

// had to change name of this so that typescript would be happy
class RandNode {
  val: number;
  next: RandNode | null;
  random: RandNode | null;
  constructor(val?: number, next?: RandNode, random?: RandNode) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
    this.random = random === undefined ? null : random;
  }
}
function copyRandomList(head: RandNode | null): RandNode | null {
  if (head === null) {
    return head;
  }

  // build interweaved list
  let current = head;
  while (current !== null) {
    current.next = new RandNode(current.val, current.next, null);
    current = current.next.next;
  }

  // set random
  current = head;
  while (current !== null) {
    if (current.random) {
      current.next.random = current.random.next;
    }
    current = current.next.next;
  }

  current = head;
  let newHead = head.next;
  while (current !== null) {
    let newNode = current.next;
    current.next = newNode.next;
    current = newNode.next;
    if (newNode.next !== null) {
      newNode.next = newNode.next.next;
    }
  }
  // fix lists
  return newHead;
}
// @lc code=end
